crates/bls12381/src/fields/fp.rs

```diff
@@ -0,0 +1,504 @@
+use bellpepper_core::{
+    boolean::{AllocatedBit, Boolean},
+    ConstraintSystem, SynthesisError,
+};
+use bellpepper_emulated::field_element::{
+    EmulatedFieldElement, EmulatedFieldParams, PseudoMersennePrime,
+};
+use bls12_381::fp::Fp as BlsFp;
+use ff::{PrimeField, PrimeFieldBits};
+use num_bigint::{BigInt, Sign};
+
+#[derive(Debug)]
+pub struct Bls12381FpParams;
+
+impl EmulatedFieldParams for Bls12381FpParams {
+    // TODO: Depending on the native field, different limb/bit pairs are more optimal and have less waste. This should be customizable and not hardcoded
+    // for example, in the pasta field, 4/96 could be used instead
+    fn num_limbs() -> usize {
+        6
+    }
+
+    fn bits_per_limb() -> usize {
+        64
+    }
+
+    fn modulus() -> BigInt {
+        BigInt::parse_bytes(
+            b"1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab",
+            16,
+        )
+        .unwrap()
+    }
+
+    fn is_modulus_pseudo_mersenne() -> bool {
+        false
+    }
+
+    fn pseudo_mersenne_params() -> Option<PseudoMersennePrime> {
+        None
+    }
+}
+
+pub type Bls12381Fp<F> = EmulatedFieldElement<F, Bls12381FpParams>;
+
+#[derive(Clone)]
+pub struct AllocatedFieldElement<F: PrimeField + PrimeFieldBits>(pub(crate) Bls12381Fp<F>);
+
+#[derive(Debug)]
+pub struct Bls12381FrParams;
+
+impl EmulatedFieldParams for Bls12381FrParams {
+    fn num_limbs() -> usize {
+        4
+    }
+
+    fn bits_per_limb() -> usize {
+        64
+    }
+
+    fn modulus() -> BigInt {
+        BigInt::parse_bytes(
+            b"73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001",
+            16,
+        )
+        .unwrap()
+    }
+
+    fn is_modulus_pseudo_mersenne() -> bool {
+        false
+    }
+
+    fn pseudo_mersenne_params() -> Option<PseudoMersennePrime> {
+        None
+    }
+}
+
+pub type Bls12381Fr<F> = EmulatedFieldElement<F, Bls12381FrParams>;
+
+impl<F> From<&BlsFp> for AllocatedFieldElement<F>
+where
+    F: PrimeField + PrimeFieldBits,
+{
+    fn from(value: &BlsFp) -> Self {
+        let bytes = value.to_bytes();
+        assert!(bytes.len() == 48);
+        let val = BigInt::from_bytes_be(Sign::Plus, &bytes);
+        Self(Bls12381Fp::<F>::from(&val))
+    }
+}
+
+pub(crate) fn bigint_to_fpelem(val: &BigInt) -> Option<BlsFp> {
+    use num_traits::Zero;
+    if val >= &Bls12381FpParams::modulus() {
+        return None;
+    }
+    let be_bytes = BigInt::to_bytes_be(val);
+    if be_bytes.0 != Sign::Plus {
+        // the sign is only non-positive if the value is exactly 0
+        assert!(val == &BigInt::zero());
+    }
+    let mut bytes: Vec<u8> = be_bytes.1;
+    assert!(bytes.len() <= 48);
+    while bytes.len() < 48 {
+        bytes.insert(0, 0);
+    } // TODO: here we pad with 0 bytes up to 48 bytes, but maybe something better could be done instead
```

Comment:
`bytes.resize(48, 0)`

Reply:
That actually doesn't work because the padding needs to be inserted at the beginning of the vec rather than the end.
I can do this however, which may or may not be more readable than the while loop I had before:
```
let bytes: Vec<u8> = bytes.splice(0..0, vec![0; 48 - bytes.len()]).collect();
```

Reply:
The insert method's complexity is linear. You probably want the following:
```
let n = bytes.len();
let bytes = bytes.into_iter().rev().chain(std::iter::repeat(0).take(48-n)).collect();
```